lib/Sema/MiscDiagnostics.cpp

```diff
@@ -4438,6 +4438,143 @@ static void diagnoseExplicitUseOfLazyVariableStorage(const Expr *E,
+        : C(dc->getASTContext()), DC(dc) {}
+
+    void tryDiagnoseComparisonWithNaN(BinaryExpr *BE) {
+      ValueDecl *comparisonDecl = nullptr;
+
+      // The == and != methods take two arguments.
+      if (BE->getArg()->getNumElements() != 2) {
+        return;
+      }
+
+      // Dig out the function the arguments are being passed to.
+      if (auto Fn = BE->getFn()) {
+        if (auto DSCE = dyn_cast<DotSyntaxCallExpr>(Fn)) {
+          comparisonDecl = DSCE->getCalledValue();
+        } else {
+          comparisonDecl = BE->getCalledValue();
+        }
+      }
+
+      // Bail out if it isn't a function.
+      if (!comparisonDecl || !isa<FuncDecl>(comparisonDecl)) {
+        return;
+      }
+
+      // We're only interested in == and != functions.
+      if (!(comparisonDecl->getBaseIdentifier().is("==") ||
+            comparisonDecl->getBaseIdentifier().is("!="))) {
+        return;
+      }
+
+      auto firstArg = BE->getArg()->getElement(0);
+      auto secondArg = BE->getArg()->getElement(1);
+
+      // == is defined in FloatingPoint stdlib protocol.
+      // != comes from Equatable which uses == internally.
+      // If both arguments conform to FloatingPoint then we're
+      // in the clear and we don't have to check anything else.
+      auto conformsToFpProto = [&](Type type) {
+        auto fpProto = C.getProtocol(KnownProtocolKind::FloatingPoint);
+        auto comparisonProtoDC =
+            comparisonDecl->getDeclContext()->getSelfProtocolDecl();
+        return !TypeChecker::conformsToProtocol(type, fpProto,
+                                                const_cast<DeclContext *>(DC))
+                    .isInvalid();
+      };
+
+      if (!conformsToFpProto(firstArg->getType()) ||
+          !conformsToFpProto(firstArg->getType())) {
+        return;
+      }
+
+      // Dig out the declarations for the arguments.
+      ValueDecl *firstVal = nullptr;
+      ValueDecl *secondVal = nullptr;
+      if (auto DRE = dyn_cast<DeclRefExpr>(firstArg)) {
+        firstVal = DRE->getDecl();
+      } else if (auto MRE = dyn_cast<MemberRefExpr>(firstArg)) {
+        firstVal = MRE->getMember().getDecl();
+      }
+
+      if (auto DRE = dyn_cast<DeclRefExpr>(secondArg)) {
+        secondVal = DRE->getDecl();
+      } else if (auto MRE = dyn_cast<MemberRefExpr>(secondArg)) {
+        secondVal = MRE->getMember().getDecl();
+      }
+
+      // One of them has to be '.nan', so if we don't have declarations
+      // for both, then bail out.
+      if (!firstVal && !secondVal) {
+        return;
+      }
+
+      // Convenience utility to check if this is a 'nan' variable.
+      auto isNanDecl = [&](ValueDecl *VD) {
+        return VD && isa<VarDecl>(VD) && VD->getBaseIdentifier().is("nan");
+      };
+
+      // Emit a special diagnostic for comparisons where both arguments is
+      // '.nan' i.e. Double.nan == Double.nan is always false and Double.nan !=
+      // Double.nan is always true.
```

Comment:
I don't think you should need to handle both-NaN specially; if _either_ is NaN, `==` is always false and `!=` is always true.

Reply:
Right, I added it because I felt suggesting "isNan" in this scenario feels a bit weird. Do you have any suggestions on what to communicate to the user if both sides are NaN?

Reply:
One option might be to spell out the decision flowchart explicitly: "To test if x is not a number, use `x.isNaN`. To test if x is a number, use `!x.isNaN`."

As an aside, we should be able to emit a diagnostic for ordered comparisons (`<`, `<=`, `=>`, `>`) as well; those are all always false, too (though less commonly seen with constant NaN arguments).

Reply:
Hmm, alright. I have added support for ordered comparisons too.

Please take a look at `test/decl/var/nan_comparisons.swift` to see test cases and associated diagnostics and let me know if it looks okay to you! The wording of the diagnostic is tailored to the kind of comparison done.